Better lua indentation

hugomg opened the pull request:
Over the last weeks I've been using Kakoune to program in Lua... I felt that some parts of the automatic indentation could be improved, so here is a PR for it :)

## Indentation

- Indent inside **do** **end**
- Indent inside **repeat** **until**
- Indent after a line ending with `{` or `(`
- Do not indent if the keyword is inside a string or comment
- More accurate un-indentation for the **end** keyword

 ```lua
-- Previously, the second "end" was
-- indented at the same level as the "if"
 while true do
      if false then
      end
      end
```

## Automatic insertion of end keyword

- Also insert after **do**, **else** and **elseif**
- Do not insert if the keyword is inside a string or comment
- Do not insert if the block already contains statements

```lua
-- No longer does this:
-- If we open a new line at the start of an unclosed block, adding
-- an "end" right after the cursor is probably not what we want.
if true then
    @
end
    x = 1
    y = 2
```

mawww commented:
This seems to break an existing test case for lua indentation, can you take a look and either update the test or fix the behaviour ?

hugomg commented:
I hadn't noticed that there were test cases for the indentation. Clever!

> can you take a look and either update the test or fix the behaviour ?

The last two commits do both things. I fixed a a bug that was caught by the test cases and I also added more test cases to document the improvements I made.

mawww commented:
Excellent, thanks for the updated version, will merge soon.
